keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1087 @@
+- [ ] User-facing documentation has been created in
+      [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+      links to mailing list discussions/SIG meetings, relevant
+      PRs/issues, release notes
+
+## Summary
+
+There are two types of volumes that are getting created after scheduling a pod onto a node:
+- [ephemeral inline
+  volumes](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
+- persistent volumes with [delayed
+  binding](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode)
+  (`WaitForFirstConsumer`)
+
+In both cases the Kubernetes scheduler currently picks a node without
+knowing whether the storage system has enough capacity left for
+creating a volume of the requested size. In the first case, `kubelet`
+will ask the CSI node service to stage the volume. Depending on the
+CSI driver, this involves creating the volume. In the second case, the
+`external-provisioner` will note that a `PVC` is now ready to be
+provisioned and ask the CSI controller service to create the volume
+such that is usable by the node (via
+[`CreateVolumeRequest.accessibility_requirements`](https://kubernetes-csi.github.io/docs/topology.html)).
+
+If these volume operations fail, pod creation may get stuck. The
+operations will get retried and might eventually succeed, for example
+because storage capacity gets freed up or extended. A pod with an
+ephemeral volume will not get rescheduled to another node. A pod with
+a volume that uses delayed binding should get scheduled multiple times,
+but then might always land on the same node unless there are multiple
+nodes with equal priority.
+
+A new API for exposing storage capacity currently available via CSI
+drivers and a scheduler enhancement that uses this information will
+reduce the risk of that happening.
+
+## Motivation
+
+### Goals
+
+* Define an API for exposing information about storage that is
+  flexible enough for a variety of use cases and that can be extended
+  later on.
+
+* Expose capacity information at the semantic
+  level that Kubernetes currently understands, i.e. in a way that
+  Kubernetes can compare capacity against the requested size of
+  volumes. This has to work for local storage, network-attached
+  storage and for drivers where the capacity depends on parameters in
+  the storage class.
+
+* Increase the chance of choosing a node for which volume creation
+  will succeed by tracking the currently available capacity available
+  through a CSI driver and using that information during pod
+  scheduling.
+
+### Non-Goals
+
+* Only CSI drivers will be supported.
+
+* No attempts will be made to model how capacity will be affected by
+  pending volume operations. This would depend on internal driver
+  details that Kubernetes doesn’t have.
+
+* Nodes are not prioritized based on how much storage they have available.
```

Comment:
Prioritizing nodes with the most available capacity could help reduce the chance of provisioning failing due to stale capacity information.